net: transport abstraction

sipa opened the pull request:
This PR furthers the P2P message serialization/deserialization abstraction introduced in #16202 and #16562, in preparation for introducing the BIP324 v2 transport (making this part of #27634). However, nothing in this PR is BIP324-specific, and it contains a number of independently useful improvements.

The overall idea is to have a single object in every `CNode` (called `m_transport`) that is responsible for converting sent messages to wire bytes, and for converting received wire bytes back to messages, while having as little as possible knowledge about this conversion process in higher-level net code. To accomplish that, there is an abstract `Transport` class with (currently) a single `V1Transport` implementation.

Structurally, the above is accomplished by:
* Merging the `TransportDeserializer` and `TransportSerializer` classes into a single `Transport` class, which encompasses both the sending and receiving side. For `V1Transport` these two sides are entirely separate, but this assumption doesn't hold for the BIP324 transport where e.g. the sending encryption key depends on the DH key negotiation data received from the other side. Merging the two means a future `V2Transport` can handle all this interaction without callers needing to be aware.
* Removing the assumption that each message is sent using a computed header followed by (unmodified) data bytes. To achieve that, the sending side of `Transport` mirrors what the receiver side does: callers can set a message to be sent, then ask what bytes must be sent out, and then allowing them to transition to the next message.
* Adding internal locks to protect the sending and receiving state of the `V1Transport` implementation. I believe these aren't strictly needed (opinions welcome) as there is no real way to use `Transport` objects in a multi-threaded fashion without some form of external synchronization (e.g. "get next bytes to send" isn't meaningful to call from multiple threads at the same time without mechanism to control the order they'll actually get sent). Still, I feel it's cleaner to make the object responsible for its own consistency (as we definitely do not want the entire object to be under a single external GUARDED_BY, as that'd prevent simultaneous sending and receiving).
* Moving the conversion of messages to bytes on the sending side from `PushMessage` to `SocketSendData`, which is needed to deal with the fact that a transport may not immediately be able to send messages.

This PR is not a refactor, though some commits are. Among the semantic changes are:
* Changing the send buffer pushback mechanism to trigger based on the memory usage of the buffer rather than the amount of bytes to be sent. This is both closer to the desired behavior, and makes the buffering independent from transport details (which is why it's included here).
* When optimistic send is not applicable, the V1 message checksum calculation now runs in the net thread rather than the message handling thread. I believe that's generally an improvement, as the message handling thread is far more computationally bottlenecked already.
* The checksum calculation now runs under the `CNode::cs_vSend` lock, which does mean no two checksum calculations for messages sent to the same node can run in parallel, even if running in separate threads. Despite that limitation, having the checksum for non-optimistic sends moved in the net thread is still an improvement, I believe.
* Statistics for per-message-type sent bytes are now updated when the bytes are actually handed to the OS rather than in `PushMessage`. This is because the actual serialized sizes aren't known until they've gone through the transport object.

A fuzz test of the entire `V1Transport` is included. More elaborate rationale for each of the changes can be found in the commit messages.

sipa committed:
refactor: merge transport serializer and deserializer into Transport class

This allows state that is shared between both directions to be encapsulated
into a single object. Specifically the v2 transport protocol introduced by
BIP324 has sending state (the encryption keys) that depends on received
messages (the DH key exchange). Having a single object for both means it can
hide logic from callers related to that key exchange and other interactions.

sipa committed:
net: abstract sending side of transport serialization further

This makes the sending side of P2P transports mirror the receiver side: caller provides
message (consisting of type and payload) to be sent, and then asks what bytes must be
sent. Once the message has been fully sent, a new message can be provided.

This removes the assumption that P2P serialization of messages follows a strict structure
of header (a function of type and payload), followed by (unmodified) payload, and instead
lets transports decide the structure themselves.

It also removes the assumption that a message must always be sent at once, or that no
bytes are even sent on the wire when there is no message. This opens the door for
supporting traffic shaping mechanisms in the future.

sipa committed:
net: measure send buffer fullness based on memory usage

This more accurately captures the intent of limiting send buffer size, as
many small messages can have a larger overhead that is not counted with the
current approach.

It also means removing the dependency on the header size (which will become
a function of the transport choice) from the send buffer calculations.

sipa committed:
fuzz: add bidirectional fragmented transport test

This adds a simulation test, with two V1Transport objects, which send messages
to each other, with sending and receiving fragmented into multiple pieces that
may be interleaved. It primarily verifies that the sending and receiving side
are compatible with each other, plus a few sanity checks.

sipa committed:
net: add V1Transport lock protecting receive state

Rather than relying on the caller to prevent concurrent calls to the
various receive-side functions of Transport, introduce a private m_cs_recv
inside the implementation to protect the lock state.

Of course, this does not remove the need for callers to synchronize calls
entirely, as it is a stateful object, and e.g. the order in which Receive(),
Complete(), and GetMessage() are called matters. It seems impossible to use
a Transport object in a meaningful way in a multi-threaded way without some
form of external synchronization, but it still feels safer to make the
transport object itself responsible for protecting its internal state.

sipa committed:
net: make V1Transport implicitly use current chainparams

The rest of net.cpp already uses Params() to determine chainparams in many
places (and even V1Transport itself does so in some places).

Since the only chainparams dependency is through the message start characters,
just store those directly in the transport.

sipa committed:
refactor: make Transport::ReceivedBytes just return success/fail

dergoegge commented:
Concept ACK

> Merging the two means a future V2Transport can handle all this interaction without callers needing to be aware.

🚀

Sjors reviewed (commented):
Concept ACK

Mostly happy with b7a7ed70d06ab3f994ff58e3a0c99105ee88ab6e. Minor inline suggestions and/or making sure I understand what the code is doing.

I also checked that all intermediate commits compile and ran the new fuzzer for several hours (but didn't study it very carefully).

* 1937a5fcf795149c44b7f4f016c05000ac3adaf9: `-maxsendbuffer` help could be changed to say "Maximum per-connection memory use for the send buffer"
* 68e48a0185751d24eecb194b8efd7028c8b590f3: can you elaborate a bit in the commit description why `nSendOffset` can be dropped?

Sjors commented on src/net.cpp:
```diff
@@ -897,37 +897,39 @@ size_t CConnman::SocketSendData(CNode& node) const
-        {
+        if (!data.empty()) {
             LOCK(node.m_sock_mutex);
             if (!node.m_sock) {
```
I guess it's not entirely clear to me whose responsibility it is to ensure the handshake has been done: https://github.com/bitcoin/bitcoin/pull/28165#discussion_r1283674053 (I assume it will be in the main PR)

Sjors replied:
I think `!node.m_sock` can only happen when when we've disconnected in another thread, using `CloseSocketDisconnect()` (and during tests). That's also what the documentation of `m_sock_mutex` says. cc @vasild

Sjors replied:
8d1a91f6394f00b880613268a9f2f3164ef914e2: technically `have_unsent_data` should be set to `true` here, but since we just disconnected it doesn't matter.

vasild replied:
Right, if `CNode::m_sock` is empty `shared_ptr`, that means `CNode::CloseSocketDisconnect()` has been called and this is a one-way street, it is not coming back, so all communication structs and data for this peer can be dropped.

Or, during tests an empty `shared_ptr` may be supplied to the `CNode` constructor. My understanding is that in those tests the code that uses the socket is not supposed to be reached/executed. Other tests provide a mocked socket, that returns either fuzzed data or static/hardcoded contents upon read in order to entertain the code that receives from the socket.

theStack commented on src/net.h:
```diff
@@ -271,10 +271,26 @@ class Transport {
+    // another.
+
+    /** Whether the last provided message has been sent, and a new one can be provided. */
+    virtual bool DoneSendingMessage() const noexcept = 0;
```
FWIW, I agree that the `Done` prefix of the method name is quite confusing, for both the reasons Sjors lined out. `CanSetNewMessageToSend()` sounds much better to me (maybe even just `CanSetMessageToSend`, to have the full name of the method included which can be called if `true` is returned?).

theStack reviewed (commented):
Concept ACK

Slowly working my way through, reviewed up to f97adbf3e93e89b1e8ce1dc212e84ac6b2879463 (commit 3/8), looks good so far.
One potential code deduplication nit (though I'm not sure if it gains that much in readability): seems like for V1Transport,  `HaveBytesToSend()` and `DoneSendingMessage()` are in an inverse relation (or however one would call that), so one could be expressed through the other, e.g.:
```diff
index 887669e32b..af0fa5c603 100644
--- a/src/net.cpp
+++ b/src/net.cpp
@@ -863,8 +863,7 @@ void V1Transport::SetMessageToSend(CSerializedNetMsg&& msg) noexcept
 
 bool V1Transport::HaveBytesToSend() const noexcept
 {
-    LOCK(m_cs_send);
-    return m_sending_header || m_bytes_sent != m_message_to_send.data.size();
+    return !DoneSendingMessage();
 }
 
 Transport::BytesToSend V1Transport::GetBytesToSend() const noexcept
```

vincenzopalazzo commented on src/net.cpp:
```diff
@@ -827,8 +834,46 @@ void V1Transport::prepareForTransport(CSerializedNetMsg& msg, std::vector<unsign
     memcpy(hdr.pchChecksum, hash.begin(), CMessageHeader::CHECKSUM_SIZE);
 
     // serialize header
-    header.reserve(CMessageHeader::HEADER_SIZE);
```
Mh looks like that `header` is no longer used?

vincenzopalazzo reviewed (commented):
Concept ACK till https://github.com/bitcoin/bitcoin/pull/28165/commits/1937a5fcf795149c44b7f4f016c05000ac3adaf9

I will continue later on it

mzumsande commented on src/test/util/net.cpp:
```diff
@@ -72,6 +72,12 @@ void ConnmanTestMsg::NodeReceiveMsgBytes(CNode& node, Span<const uint8_t> msg_by
 bool ConnmanTestMsg::ReceiveMsgFrom(CNode& node, CSerializedNetMsg&& ser_msg) const
 {
+    /* Flush out any unsent bytes from previous messages. */
+    while (node.m_transport->HaveBytesToSend()) {
```
It seems confusing and not ideal that a function called `ReceiveMsgFrom` has so much code dealing with the Send part, and also the side effect of flushing the send message buffer, when the goal is just to create a header for `ser_msg` to be able to receive that, but not to send anything. For example, it should be possible to call `ReceiveMsgFrom` in situations where the send buffer has unrelated contents. Maybe it'd be better to just have the relevant code (i.e. the old `prepareForTransport` duplicated here to extract the header instead of changing the send parts of `m_transport`?)

mzumsande reviewed (commented):
Concept ACK

sipa committed:
refactor: rename Transport class receive functions

Now that the Transport class deals with both the sending and receiving side
of things, make the receive side have function names that clearly indicate
they're about receiving.

* Transport::Read() -> Transport::ReceivedBytes()
* Transport::Complete() -> Transport::ReceivedMessageComplete()
* Transport::GetMessage() -> Transport::GetReceivedMessage()
* Transport::SetVersion() -> Transport::SetReceiveVersion()

Further, also update the comments on these functions to (among others) remove
the "deserialization" terminology. That term is better reserved for just the
serialization/deserialization between objects and bytes (see serialize.h), and
not the conversion from/to wire bytes as performed by the Transport.

Sjors reviewed (commented):
The new approach in 384c798cc3836558463e88ec7f563b236f50bf22 is a nice improvement.

I'll continue reviewing from 09972b518bed5809d37c79fb0ddef034dce27ba0 onwards after you've had a chance to comment on my `more` comment.

sipa committed:
net: move message conversion to wire bytes from PushMessage to SocketSendData

This furthers transport abstraction by removing the assumption that a message
can always immediately be converted to wire bytes. This assumption does not hold
for the v2 transport proposed by BIP324, as no messages can be sent before the
handshake completes.

This is done by only keeping (complete) CSerializedNetMsg objects in vSendMsg,
rather than the resulting bytes (for header and payload) that need to be sent.
In SocketSendData, these objects are handed to the transport as permitted by it,
and sending out the bytes the transport tells us to send. This also removes the
nSendOffset member variable in CNode, as keeping track of how much has been sent
is now a responsability of the transport.

This is not a pure refactor, and has the following effects even for the current
v1 transport:

* Checksum calculation now happens in SocketSendData rather than PushMessage.
  For non-optimistic-send messages, that means this computation now happens in
  the network thread rather than the message handler thread (generally a good
  thing, as the message handler thread is more of a computational bottleneck).
* Checksum calculation now happens while holding the cs_vSend lock. This is
  technically unnecessary for the v1 transport, as messages are encoded
  independent from one another, but is untenable for the v2 transport anyway.
* Statistics updates about per-message sent bytes now happen when those bytes
  are actually handed to the OS, rather than at PushMessage time.

Sjors reviewed (approved):
ACK d9a91ba9606f828ef3846d2877ec56313653a109 just in time for the #27981 rebase :-(

Sjors commented:
re-ACK 07cb6c4

The rebase on #27981 looks good to me. I didn't look at the fuzzer changes. Otherwise it's just some extra comments.

Sjors commented:
> * (sorry @Sjors, the earlier rebase was incorrect, it only looked at `vSendMsg.empty()` for the returned `data_left` in `SocketSendData()`).

Is there a test you can add that would catch this difference?

Sjors commented on src/net.cpp:
```diff
@@ -906,36 +906,50 @@ std::pair<size_t, bool> CConnman::SocketSendData(CNode& node) const
-    while (it != node.vSendMsg.end()) {
-        const auto& data = *it;
-        assert(data.size() > node.nSendOffset);
+    bool have_unsent_data{false};
```
8d1a91f6394f00b880613268a9f2f3164ef914e2: `data_left` is more consistent with the header

Sjors replied:
Or perhaps something like `return {nSentSize, socket_empty && message_queue_empty};`

Sjors commented on src/net.cpp:
```diff
@@ -906,36 +906,50 @@ std::pair<size_t, bool> CConnman::SocketSendData(CNode& node) const
+                // Update memory usage of send buffer (as *it will be deleted).
+                node.m_send_memusage -= memusage;
+                ++it;
+            }
```
8d1a91f6394f00b880613268a9f2f3164ef914e2: IIUC even though `!node.vSendMsg.empty()`, we don't want `else { have_unsent_data = true }` here because:

1. `SetMessageToSend` _may_ return `false` for other reasons (though afaik it can't at the moment)
2. If it does return `false`, then `GetBytesToSend` below populates `data`, tries to send it and then updates `have_unsent_data` appropriately.

Sjors reviewed (approved):
re-ACK d99269058922af5fdc14d7fcc88f132060b4b57a

vincenzopalazzo reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/28165/commits/d99269058922af5fdc14d7fcc88f132060b4b57a

I like the abstraction of the Transport concept, it is close to a person's mental model of this kind of system!

vasild commented on src/test/fuzz/p2p_transport_serialization.cpp:
```diff
@@ -63,22 +65,260 @@ FUZZ_TARGET(p2p_transport_serialization, .init = initialize_p2p_transport_serial
+    };
+
+    // Main loop, interleaving new messages, sends, and receives.
+    unsigned iter = 0;
+    while (iter < 1000 && provider.remaining_bytes()) {
```
```suggestion
    LIMITED_WHILE(provider.remaining_bytes(), 1000) {
```

vasild commented on src/test/fuzz/p2p_transport_serialization.cpp:
```diff
@@ -63,22 +65,260 @@ FUZZ_TARGET(p2p_transport_serialization, .init = initialize_p2p_transport_serial
+    while (iter < 1000 && provider.remaining_bytes()) {
+        uint8_t code = provider.ConsumeIntegral<uint8_t>();
+        // Lowest bit identifies the sender side of the operation.
+        int side = code & 1;
+        switch ((side >> 1) % 3) {
```
Given that `side` is `0` or `1`, then `side >> 1` will always be `0`, no? Did you mean `code >> 1`? I find this less cryptic:

```cpp
size_t side = provider.ConsumeBool();
switch (provider.ConsumeIntegralInRange(0, 2)) {
    case 0: ...
    case 1: ...
    case 2: ...
}

vasild commented on src/test/fuzz/p2p_transport_serialization.cpp:
```diff
@@ -63,22 +65,260 @@ FUZZ_TARGET(p2p_transport_serialization, .init = initialize_p2p_transport_serial
+    // Test with two V1 transports talking to each other.
+    FuzzedDataProvider provider{buffer.data(), buffer.size()};
+    XoRoShiRo128PlusPlus rng(provider.ConsumeIntegral<uint64_t>());
+    auto t1 = MakeV1Transport(NodeId{0});
+    auto t2 = MakeV1Transport(NodeId{1});
+    if (!t1 || !t2) return;
+    SimulationTest(*t1, *t2, rng, provider);
```
Is there a reason to use pointers? The following seems to be simpler:

```cpp
    V1Transport t0{NodeId{0}, SER_NETWORK, INIT_PROTO_VERSION};
    V1Transport t1{NodeId{1}, SER_NETWORK, INIT_PROTO_VERSION};
    SimulationTest(t0, t1, rng, provider);
```
and remove `MakeV1Transport()`?

vasild replied:
It is ok as it is - there is _some_ reason for using pointers.

vasild commented on src/test/fuzz/p2p_transport_serialization.cpp:
```diff
@@ -63,22 +65,260 @@ FUZZ_TARGET(p2p_transport_serialization, .init = initialize_p2p_transport_serial
+            // Construct a new next message to send.
+            next_msg[side] = make_msg_fn(false);
         }
+        // Return whether any message was added.
+        return queued;
```
The return value of `new_msg_fn()` is ignored by the caller.
```suggestion
```

vasild reviewed (commented):
Posting review midway, reviewed up to 57e9bb0ffd (incl).

vasild commented on src/net.h:
```diff
@@ -253,42 +256,92 @@ class CNetMessage {
+    virtual bool ReceivedMessageComplete() const = 0;
+    // set the deserialization context version
+    virtual void SetReceiveVersion(int version) = 0;
+    /** read and deserialize data, advances msg_bytes data pointer. Return false is invalid. */
+    virtual bool ReceivedBytes(Span<const uint8_t>& msg_bytes) = 0;
```
`s/false is invalid/false if invalid/` or even better, tell doxygen about it:

```suggestion
    /**
     * Read and deserialize data, advance msg_bytes data pointer.
     * @return false if invalid.
     */
    virtual bool ReceivedBytes(Span<const uint8_t>& msg_bytes) = 0;
```

vasild commented on src/net.cpp:
```diff
@@ -681,16 +690,16 @@ bool CNode::ReceiveMsgBytes(Span<const uint8_t> msg_bytes, bool& complete)
         // absorb network data
-        int handled = m_deserializer->Read(msg_bytes);
-        if (handled < 0) {
+        bool handled = m_transport->ReceivedBytes(msg_bytes);
+        if (!handled) {
```
nit (`handled` is not used afterwards):
```suggestion
        if (!m_transport->ReceivedBytes(msg_bytes)) {
```
(same in `p2p_transport_serialization.cpp`)

Sjors replied:
I don't mind keeping `handled` around, to make it more clear that receiving 0 bytes is also "handled".

theStack replied:
nit: if `handled` is decided to be kept, it should be declared as `const`.

vasild reviewed (approved):
ACK 678809e00178b33e273f0775eea9635fecaf386a

Some things may be improved, see comments below, but no blockers (thus ACK).

Sjors commented on src/net.cpp:
```diff
@@ -1227,7 +1322,14 @@ Sock::EventsPerSock CConnman::GenerateWaitSockets(Span<CNode* const> nodes)
+            LOCK(pnode->cs_vSend);
+            // Sending is possible if either there are bytes to send right now, or if there will be
+            // once a potential message from vSendMsg is handed to the transport.
+            const auto& [to_send, _more, _msg_type] = pnode->m_transport->GetBytesToSend();
```
Optionally annotate the return values if the fact that we're ignoring them is potentially interesting / strange. `[to_send, /*more=*/std::ignore, /*msg_type=*/std::ignore] = pnode->…`

vasild replied:
Ok, in C++27 then, sorry for the noise.

Sjors commented on src/net.cpp:
```diff
@@ -819,54 +842,124 @@ CNetMessage V1TransportDeserializer::GetMessage(const std::chrono::microseconds
-            } else {
+            if ((size_t)nBytes != data.size()) {
                 // could not send full message; stop sending more
+                data_left = true;
```
I would use `data_in_transport` instead of `_data_left` (see https://github.com/bitcoin/bitcoin/pull/28165#discussion_r1298717971).

And then drop `data_left = _data_left;`. 

Then below:

```cpp
bool message_queue_empty = it == node.vSendMsg.end()`
if (data_in_transport || !message_queue_empty) {
  flags |= MSG_MORE;
```

And:

```
// couldn't send anything at all
if (!data.empty()) {
  data_in_transport = true
}
```

And then finally return `!message_queue_empty || data_in_transport` instead of `data_left`.

Sjors replied:
I'm only seeing an _extra_ line here…? But something along those lines could be worth a shot.

Sjors commented:
re-ACK 678809e00178b33e273f0775eea9635fecaf386a

theStack commented on src/net.cpp:
```diff
@@ -2910,27 +2957,40 @@ void CConnman::PushMessage(CNode* pnode, CSerializedNetMsg&& msg)
+        // In the current transport (V1Transport), GetBytesToSend first returns a header to send,
+        // and then the payload data (if any), necessitating a loop.
+        while (true) {
+            const auto& [bytes, more, msg_type] = pnode->m_transport->GetBytesToSend();
```
tiny-indicate-unused-nit:
```suggestion
            const auto& [bytes, _more, msg_type] = pnode->m_transport->GetBytesToSend();
```
(probably not worth retouching as that code is replaced in a later commit anyways)

vincenzopalazzo reviewed (approved):
re-ACK https://github.com/bitcoin/bitcoin/pull/28165/commits/41806a45d362636c6ffea6bdf39620c785b0edd8

Sjors reviewed (approved):
re-ACK 41806a45d362636c6ffea6bdf39620c785b0edd8

Sjors commented on src/test/fuzz/p2p_transport_serialization.cpp:
```diff
@@ -92,3 +102,231 @@ FUZZ_TARGET(p2p_transport_serialization, .init = initialize_p2p_transport_serial
+    // Two vectors representing in-flight bytes. inflight[i] is from transport[i] to transport[!i].
+    std::array<std::vector<uint8_t>, 2> in_flight;
+
+    // Two deques with expected messages. expected[i] is expected to arrive in transport[!i].
```
nit: dequeues (or add spelling exception)

theStack reviewed (approved):
Code-review ACK 41806a45d362636c6ffea6bdf39620c785b0edd8

mzumsande reviewed (commented):
ACK 41806a45d362636c6ffea6bdf39620c785b0edd8

I reviewed the code (the added fuzz test only superficially), and ran a node with this for some time and didn't encounter any problems..

theStack reviewed (approved):
re-ACK 8a3b6f33873a1075f932f5d9feb6d82e50d83c0c

The only changes since my previous ACK are two nits tackled regarding spelling (s/deques/queues/) and named arguments in the simulation test.

vasild reviewed (approved):
ACK 8a3b6f33873a1075f932f5d9feb6d82e50d83c0c

dergoegge reviewed (approved):
Code review ACK 8a3b6f33873a1075f932f5d9feb6d82e50d83c0c

vasild commented on src/test/fuzz/p2p_transport_serialization.cpp:
```diff
@@ -17,16 +19,25 @@
 {
     SelectParams(ChainType::REGTEST);
+    g_all_messages = getAllNetMessageTypes();
+    std::sort(g_all_messages.begin(), g_all_messages.end());
```
Why this sort here? Anybody remembers? It is only used like:

```cpp
g_all_messages[v % g_all_messages.size()]
```

where `v` is a random number. The sort seems unnecessary?

The question arised in: https://github.com/bitcoin/bitcoin/pull/29421#discussion_r1490653235
